### PR TITLE
[SPARK-9479][Streaming][Tests]Fix ReceiverTrackerSuite failure for maven build and other potential test failures in Streaming

### PR DESCRIPTION
**Pull request by zsxwing**
See https://issues.apache.org/jira/browse/SPARK-9479 for the failure cause.

The PR includes the following changes:
1. Make ReceiverTrackerSuite create StreamingContext in the test body.
2. Fix places that don't stop StreamingContext. I verified no SparkContext was stopped in the shutdown hook locally after this fix.
3. Fix an issue that `ReceiverTracker.endpoint` may be null.
4. Make sure stopping SparkContext in non-main thread won't fail other tests.

### CODE REVIEW

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala [hunk 1]
```diff
@@ -223,7 +223,11 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
     // Signal the receivers to delete old block data
     if (WriteAheadLogUtils.enableReceiverLog(ssc.conf)) {
       logInfo(s"Cleanup old received batch data: $cleanupThreshTime")
-      endpoint.send(CleanupOldBlocks(cleanupThreshTime))
```
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala]: I saw NullPointerException once when I ran the tests a lot of times. The `JobScheduler.eventLoop` is stopped after `ReceiverTracker.stop` and `ReceiverTracker.stop` sets `endpoint` to null. Therefore it's possible that this method is called between `ReceiverTracker.stop` and `JobScheduler.eventLoop.stop` and will throw NullPointerException.

**Comment by zsxwing** [comment]
cc @tdas

**Comment by zsxwing** [comment]
retest this please

File: streaming/src/test/scala/org/apache/spark/streaming/BasicOperationsSuite.scala [hunk 2]
```diff
@@ -81,39 +81,41 @@ class BasicOperationsSuite extends TestSuiteBase {
   test("repartition (more partitions)") {
     val input = Seq(1 to 100, 101 to 200, 201 to 300)
     val operation = (r: DStream[Int]) => r.repartition(5)
-    val ssc = setupStreams(input, operation, 2)
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/BasicOperationsSuite.scala]: Much much needed cleanup of tests. Thank you very much @zsxwing !

**Comment by tdas** [comment]
LGTM. This is awesome! AWESOME! Much needed cleanup. :)

**Comment by tdas** [comment]
I am merging this. Thanks so much @zsxwing !

**Pull request closed without merging**